Spigot-Server-Patches/0230-Max-exp-value-when-merging-orbs.patch

```diff
@@ -0,0 +1,60 @@
+From 14f0b9722752f75b6bd7e9443b5ac1f4bcaa4d68 Mon Sep 17 00:00:00 2001
+From: BillyGalbreath <Blake.Galbreath@GMail.com>
+Date: Mon, 31 Jul 2017 01:35:54 -0500
+Subject: [PATCH] Max exp value when merging orbs
+
+
+diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+index 54d081fd..18e4581b 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+@@ -428,4 +428,10 @@ public class PaperWorldConfig {
+         disableCreeperLingeringEffect = getBoolean("disable-creeper-lingering-effect", false);
+         log("Creeper lingering effect: " + disableCreeperLingeringEffect);
+     }
++
++    public int expMergeMaxValue;
++    private void expMergeMaxValue() {
++        expMergeMaxValue = getInt("exp-merge-max-value", 10);
++        log("Experience Merge Max Value: " + expMergeMaxValue);
++    }
+ }
+diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
+index 9dfc2b4d..384592f8 100644
+--- a/src/main/java/net/minecraft/server/World.java
++++ b/src/main/java/net/minecraft/server/World.java
+@@ -1146,16 +1146,25 @@ public abstract class World implements IBlockAccess {
+             EntityExperienceOrb xp = (EntityExperienceOrb) entity;
+             double radius = spigotConfig.expMerge;
+             if (radius > 0) {
++                // Paper start
++                int maxValue = paperConfig.expMergeMaxValue;
+                 List<Entity> entities = this.getEntities(entity, entity.getBoundingBox().grow(radius, radius, radius));
+                 for (Entity e : entities) {
+-                    if (e instanceof EntityExperienceOrb) {
+-                        EntityExperienceOrb loopItem = (EntityExperienceOrb) e;
+-                        if (!loopItem.dead) {
+-                            xp.value += loopItem.value;
+-                            loopItem.die();
+-                        }
++                    if (!(e instanceof EntityExperienceOrb)) {
++                        continue;
++                    }
++                    EntityExperienceOrb loopItem = (EntityExperienceOrb) e;
++                    if (loopItem.dead || loopItem.value >= maxValue) {
++                        continue;
++                    }
++                    xp.value += loopItem.value;
++                    if (xp.value > maxValue) {
++                        loopItem.value = xp.value - maxValue;
++                        continue;
```

Comment:
I'll look it over again, but the goal here is to not break the loop. It is imperative the loops cycles every orb entity or the goal is never reached at all and orb merging would have fallen short after only merging the first few orbs.

What my goal is: Lets say 100 is your max value. You kill the dragon and it drops out 1000 xp orbs all worth 6-10 xp each. Instead of all these orbs to merge into one giant orb worth 6k-10k I want to make 60-100 orbs worth 100 xp each. This allows for the beautiful explosion/shower of exp orbs, while still keeping the entity count low enough to prevent lag.

I will double check my logic, but I believe I am spot on and have been using this on my production server for half a month now.